docs(metadat): bundle metadata shape #2 @W-7263561@ (#34)
salesforce/lwc-rfcs: text/0000-canonical-metadata.md

```
</details>

## Metadata Shape Option 2

The main differences in proposed shape #2:
- the root object interface:  not only the root object contains metadata for each file, it also describes the shape of an overall bundle/component via interface object, which contains public properties, slots, events or anything that is accessible from the outside of the component/module. The reason for doing so is that individual file metadata doesn't have the notion of other files and therefore cannot make assumption of whether it is a component. 
- no decorators
- file metadata revolves around declarations (both internal and external). The reason for this is that we don't know the content of the file, thus the metadata is collected without any assumptions - could be module with few _const_ exports, could be a class, etc.

> 

Reviewer: Add that this shape has diagnostics.

### TypeScript

#### Root Metadata Object
```ts
interface BundleMetadata {
    success: boolean;
    version: string;
    diagnostics: Array<Diagnostic>;
    entry: string,
    result: { [name: string]: FileMetadata } // metadata per each source

>     [api or interface]: ComponentInterface, // public methods, events, css properties, exports, slots.

Reviewer: Document that this is a simpler way for an external user to see the public API snapshot of a given bundle without having to drill down and deduce it.

}

> interface BundleInterface {

Reviewer: ```suggestion
interface ComponentInterface {
```

    type: 'module' | 'component',
    attachedEvents: Array<Event>,
    emittedEvents: Array<Event>,
    publicProperties: Array<ClassMember>,
    publicMethods: Array<ClassMember>,
    slots: Array<Slot>,
    exports: Array<ClassMetadata | FunctionMetadata | VariableMetadata>
}
```

<details>
<summary>Click to view format</summary>
#### Root Metadata Object Per File
```ts
export interface FileMetadata {
    path: string,
    type: 'module' | 'component'| 'css',
    dependencies: Array<Reference>, // dynamic import is treated as a module dependency type of 'dynamic'
}
```

For every source in the bundle there will be a corresponding typed class which extends from the base FileMetadata.

#### Script Metadata (.js|.ts)
```ts
interface ScriptMetadata extends FileMetadata {
    declarations: Array<ClassMetadata | FunctionMetadata | VariableMetadata>
} 

interface ClassMetadata {
   name: string,
   extends: ExtendsMeta, // id, resource, location of the super 
   classMembers: Array<ClassMember>, // includes private/public props/methods
   documentation: ClassDocumentation,

>    events: Array<ClassEvent>,

Reviewer: Add listeners

}

interface ClassDocumentation {
    classDescription: string,
    html: string,
    metadata: Object,
}

interface ExtendsMeta {
    id: string,
    type: Class | Function
    resource: string,
    location: Location,
}

interface ClassMember {
    name: string,
    type: string,
    isPublic: boolean,
    documentation?: ClassMemberDocumentation,
    wire?: WireDependency
}

interface ClassMemberDocumentation {
    description: string,
    defaultValue: string,
    dataType: string,
    isRequired: boolean
}

> interface ClassProperty {

Reviewer: ```suggestion
interface ClassProperty extends ClassMember {
```

    hasGetter: boolean,
    hasSetter: boolean,
    value?: ClassMemberValue,
}

> interface ClassMethod {

Reviewer: ```suggestion
interface ClassMethod extends ClassMemver {
```

>     hasGetter: boolean,
>     hasSetter: boolean,

Reviewer: Not required
```suggestion
    hasGetter: boolean,
    hasSetter: boolean,
```

    returnValue?: ClassMemberValue,
}

interface ClassMemberValue {
    type: ClassMemberValueType, 
    value: any,
    importedName: string,
}

enum ClassMemberValueType {
    ARRAY, BOOLEAN, MODULE, NUMBER, NULL, OBJECT, STRING, UNDEFINED, UNRESOLVED,
}

interface WireDependency {
    adapter: WireTargetAdapter;
    params?: { [name: string]: Value }; // value has type and actual value
    
}

interface WireTargetAdapter {

>     name: string;

Reviewer: ```suggestion
    adapterId: string;
```

>     reference: string;

Reviewer: What is the reference points to a wire adapter in the same module.

>     reference: string;

Reviewer: ```suggestion
    moduleSpecifier: string;
```

}
```

### HTML Metadata
```ts

> interface ComponentMetadata implements FileMetadata {

Reviewer: Is this for the root HTML file or for an individual component referenced in the template?

    attributes: Array<HTMLAttribute>,
    properties: Array<HTMLProperty>,
    slots: Array<Slot>
    staticResources: Array<HTMLReference>, //images, urls, etc with location
    events: Array<HTMLEvent>, // not sure if the even belongs to a js or html meta
}
```

### CSS Metadata 
```ts

> interface type HTMLMetadata implements FileMetadata {

Reviewer: ```suggestion
interface type CSSMetadata implements FileMetadata {
```

```
</details>

## Metadata Shape Option 2
Overview:
The vision is to allow expert users to retrieve file based metadata and for the regular users to retrieve an easy, self-contained metadata object that exposes all publicly available/accessible properties of the current bundle.

The main differences in proposed shape #2:

> - no decorators: external consumer shouldn't be concerned if the class fields are decorated or not. All they need is the information that describes the component's/module's public interface. With that, instead of collecting decorators, each class member will carry 'isPublic' property to indicate its access. 

Author: ```suggestion
- no decorators: external consumer shouldn't be concerned if the class fields are decorated or not. All they need is the information that describes whether the class member is public. With that, instead of collecting decorators, each class member will carry 'isPublic' property to indicate its access. 
```
```suggestion
- no decorators: external consumer shouldn't be concerned if the class fields are decorated or not. All they need is the information that describes the component's/module's public interface. With that, instead of collecting decorators, each class member will carry 'isPublic' property to indicate its access. 
```
```suggestion
- no decorators: external consumer shouldn't be concerned if the class fields are decorated or not. All they need is the information that describes whether the class member is public. With that, instead of collecting decorators, each class member will carry 'isPublic' property to indicate its access. 
```

- bundle interface: an interpretation of the bundle shape. It is an object that contains public properties, slots, events or anything that is accessible from the outside of the component/module. The reason for doing so is that individual file metadata doesn't have the notion of other files in the bundle and therefore cannot make assumption about an overall shape. With that, the interface is an additional step that attempts to deduce the shape and provide a union of all public properties. 

### TypeScript

#### Root Metadata Object
```ts
interface BundleMetadata {
    success: boolean;
    version: string;
    diagnostics: Array<Diagnostic>;
    entry: string,
    results: { [name: string]: FileMetadata } // metadata for each source
    [api or interface]: BundleInterface, // public methods, events, css properties, exports, slots.
}

// A bundle interface, which contains a union of all file-based
// metadata objects with some wits about inheritance and overrides. 
interface BundleInterface {
    type: 'module' | 'component',

>     attachedEvents: Array<Event>,

Author: ```suggestion
    attachedEvents: Array<Event>,
```
```suggestion
    eventListener: Array<EventListeners>,
```

    emittedEvents: Array<Event>,
    publicProperties: Array<ClassMember>,
    publicMethods: Array<ClassMember>,
    slots: Array<Slot>,
    exports: Array<ClassMetadata | FunctionMetadata | VariableMetadata>
}
```

<details>
<summary>Click to view format</summary>
#### Root Metadata Object Per File
```ts
export interface FileMetadata {

>     path: string,

Author: ```suggestion
    path: string,
    name: string,
```

>     type: 'module' | 'component'| 'css',

Author: ```suggestion
    type: 'module' | 'html'| 'css',
```

    dependencies: Array<Reference>, // dynamic import is treated as a module dependency type of 'dynamic'
}
```

For every source in the bundle there will be a corresponding typed class which extends from the base FileMetadata.

#### Script Metadata (.js|.ts)
```ts
interface ScriptMetadata extends FileMetadata {

>     fileType: string,

Author: ```suggestion
    type: 'script',
```

>     fileName: string,

Author: ```suggestion
```

    declarations: Array<ClassMetadata | FunctionMetadata | VariableMetadata>
} 

interface ClassMetadata {

>    name: string,

Author: ```suggestion
   name?: string, // optional due to anonymous class
```

   extends: ExtendsMeta, // id, resource, location of the super 
   classMembers: Array<ClassMember>, // includes private/public props/methods
   documentation: ClassDocumentation,
   events: Array<ClassEvent>,
   listeners: Array<ClassListener> // programmatic or declarative event listeners found in the file
}

interface ClassDocumentation {
    classDescription: string,
    html: string,
    metadata: Object,
}

interface ExtendsMeta {
    id: string,
    type: Class | Function
    resource: string,
    location: Location,
}

interface ClassMember {
    name: string,
    type: string,
    isPublic: boolean,
    documentation?: ClassMemberDocumentation,
    wire?: WireDependency
}

interface ClassMemberDocumentation {
    description: string,
    defaultValue: string,
    dataType: string,
    isRequired: boolean
}

> interface ClassProperty extends ClassMember{

Author: ```suggestion
interface ClassProperty extends ClassMember {
```

    hasGetter: boolean,
    hasSetter: boolean,
    value?: ClassMemberValue,
}

interface ClassMethod extends ClassMember {
    returnValue?: ClassMemberValue,
}

interface ClassMemberValue {
    type: ClassMemberValueType, 
    value: any,
    importedName: string,
}

enum ClassMemberValueType {
    ARRAY, BOOLEAN, MODULE, NUMBER, NULL, OBJECT, STRING, UNDEFINED, UNRESOLVED,
}

interface WireDependency {
    adapter: WireTargetAdapter;
    params?: { [name: string]: Value }; // value has type and actual value
    
}

interface WireTargetAdapter {
    adapterId: string;
    moduleSpecifier: string; // points to the adapter resource
}
```

### HTML Metadata
```ts
interface HTMLMetadata implements FileMetadata {

>     fileType: string,

Author: ```suggestion
```
```suggestion
    type: 'html',
```

>     fileName: string,

Author: ```suggestion
```

    tags: Array<CustomElementMetadata>
    slots: Array<Slot>

>     staticResources: Array<HTMLReference>, //images, urls, etc with location

Author: ```suggestion
    staticResources: Array<HTMLReference>, // images, urls, etc with location
```

}

> interface CustomElementMetadata {

Author: ```suggestion
interface CustomElementMetadata {
 slots: Array<Slot>
```

    attributes: Array<HTMLAttribute>,
    properties: Array<HTMLProperty>,

>     events: Array<HTMLEvent>, // not sure if the even belongs to a js or html meta

Author: ```suggestion
    listeners: Array<EventListener>, // not sure if the even belongs to a js or html meta
```

}
```

### CSS Metadata 
```ts
interface type CSSMetadata implements FileMetadata {

>     fileType: string,

Author: ```suggestion
    type:'css',
```

>     fileName: string,

Author: ```suggestion
```

    imports: Array<Reference>, // css only module imports
    customProperties: Array<CssCustomProperty>,
}

interface CssCustomProperty {
    external: boolean, // determines if the custom property is defined in the current file
    name: string,
    fallback: string,
    value?: {
        type: string,
        value: string,

>         fallback: string

Author: ```suggestion
        fallback: string | CssCustomProperty
```
